src/main/java/com/extendedclip/papi/expansion/player/PlayerExpansion.java

```diff
@@ -1,270 +1,293 @@
+        high = this.getString("ping_color.high", "&a");
+        medium = this.getString("ping_color.medium", "&e");
+        low = this.getString("ping_color.low", "&c");
+
+        if (identifier.startsWith("ping_")) {
+            identifier = identifier.split("ping_")[1];
+            Player t = Bukkit.getPlayer(identifier);
+            if (t != null) {
+                return getPing(t);
+            }
+            return "0";
+        }
+        if (identifier.startsWith("colored_ping_")) {
+            identifier = identifier.split("colored_ping_")[1];
+            Player t = Bukkit.getPlayer(identifier);
+            if (t != null) {
+                int p = Integer.valueOf(getPing(t));
+                return ChatColor.translateAlternateColorCodes('&', p > 100 ? low : p >= 50 ? medium : high) + getPing(t);
+            }
+        }
+
+        if (player == null) {
+            return "";
+        }
+
+        // offline placeholders
+        switch (identifier) {
+            case "name":
+                return player.getName();
+            case "uuid":
+                return player.getUniqueId().toString();
+            case "has_played_before":
+                return bool(player.hasPlayedBefore());
+            case "online":
+                return bool(player.isOnline());
+            case "is_whitelisted":
+                return bool(player.isWhitelisted());
+            case "is_banned":
+                return bool(player.isBanned());
+            case "is_op":
+                return bool(player.isOp());
+            case "first_played":
+            case "first_join":
+                return String.valueOf(player.getFirstPlayed());
+            case "first_played_formatted":
+            case "first_join_date":
+                return PlaceholderAPIPlugin.getDateFormat().format(new Date(player.getFirstPlayed()));
+            case "last_played":
+            case "last_join":
+                return String.valueOf(player.getLastPlayed());
+            case "last_played_formatted":
+            case "last_join_date":
+                return PlaceholderAPIPlugin.getDateFormat().format(new Date(player.getLastPlayed()));
+            case "bed_x":
+                return player.getBedSpawnLocation() != null ? String.valueOf(player.getBedSpawnLocation().getX()) : "";
+            case "bed_y":
+                return player.getBedSpawnLocation() != null ? String.valueOf(player.getBedSpawnLocation().getY()) : "";
+            case "bed_z":
+                return player.getBedSpawnLocation() != null ? String.valueOf(player.getBedSpawnLocation().getZ()) : "";
+            case "bed_world":
+                return player.getBedSpawnLocation() != null ? player.getBedSpawnLocation().getWorld()
+                        .getName() : "";
+        }
+
+        // online placeholders
+        if (!player.isOnline()) {
+            return "";
+        }
+
+        Player p = player.getPlayer();
```

Comment:
Is that one actually required for a OfflinePlayer?
Does it not just act as a normal player when they are online?

Reply:
OfflinePlayer isnt Player
if u want to use OffilePlayer as online ull have to add `.getPlayer` after OfflinePlayer
so that makes it easier instead of adding `.getPlayer` every time